examples/rcn.py

```diff
@@ -0,0 +1,392 @@
+
+# %% [markdown]
+# ### 3.2.2 Make the factor graph
+
+# %%
+start = end
+fg = graph.FactorGraph(variables=variables_all_models)
+for idx in range(edges.shape[0]):
+    edge = edges[idx]
+
+    for e in edge:
+        i1, i2, r = e
+        fg.add_factor(
+            [(idx, i1), (idx, i2)],
+            phis[r],
+            np.zeros(phis[r].shape[0]),
+        )
+
+end = time.time()
+print(f"Creating factors took {end-start:.3f} seconds.")
+
+
+# %% [markdown]
+# # 4. Run inference
+
+# %% [markdown]
+# ## 4.1 Helper functions to initialize the evidence for a given image
+
+# %%
+def get_bu_msg(img: np.ndarray) -> np.ndarray:
+    """Computes the bottom-up messages given a test image.
+    Args:
+        img: The rgb image to compute bottom up messages on (H x W x 3).
+    Returns:
+        bu_msg: An array of shape [16 x H x W] denoting the presence or absence of an oriented line segment at a particular location.
+                The elements of this array belong to the set {+1, -1}.
+    """
+
+    num_orients = 16
+    brightness_diff_threshold = 40.0
+
+    filtered = np.zeros((filters.shape[0],) + img.shape, dtype=np.float32)
+    for i in range(filters.shape[0]):
+        kern = filters[i, :, :]
+        filtered[i] = fftconvolve(img, kern, mode="same")
+
+    localized = np.zeros_like(filtered)
+    cross_orient_max = filtered.max(0)
+    filtered[filtered < 0] = 0
+    for i, (layer, suppress_mask) in enumerate(zip(filtered, suppression_masks)):
+        competitor_maxs = maximum_filter(layer, footprint=suppress_mask, mode="nearest")
+        localized[i] = competitor_maxs <= layer
+    localized[cross_orient_max > filtered] = 0
+
+    # Threshold and binarize
+    localized *= (filtered / brightness_diff_threshold).clip(0, 1)
+    localized[localized < 1] = 0
+
+    pooled_channel_weights = [(0, 1), (-1, 1), (1, 1)]
+    pooled_channels = [-np.ones_like(sf) for sf in localized]
+    for i, pc in enumerate(pooled_channels):
+        for channel_offset, factor in pooled_channel_weights:
+            ch = (i + channel_offset) % num_orients
+            pos_chan = localized[ch]
+            if factor != 1:
+                pos_chan[pos_chan > 0] *= factor
+            np.maximum(pc, pos_chan, pc)
+    bu_msg = np.array(pooled_channels)
+    return bu_msg
+
+
+# %%
+def initialize_evidences(test_img: np.ndarray) -> Dict:
+    """Computes the initial evidences to the PGMax factor graph given a test image.
+    Args:
+        test_img: The image to run inference on.
+    Returns:
+        evidence_updates: A dictionary containing the initial messages to all the variables in the factor graph.
+    """
+
+    bu_msg = get_bu_msg(test_img)
+
+    evidence_updates = {}
+    for idx in range(frcs.shape[0]):
+        frc = frcs[idx]
+
+        unary_msg = -1 + np.zeros((frc.shape[0], M))
+        for v in range(frc.shape[0]):
+            f, r, c = frc[v, :]
+            evidence = bu_msg[f, r - hps : r + hps + 1, c - hps : c + hps + 1]
```

Comment:
You are using `hps` twice here. It looks like a bug. A harmless one, but still a bug.